test_runner/regress/test_tenant_detach.py

```diff
@@ -108,6 +108,218 @@ def test_tenant_reattach(
         assert env.pageserver.log_contains(".*download.*failed, will retry.*")
 
 
+# Detach and re-attach a tenant with a branch
+#
+# This exercises a race condition after tenant attach, where the
+# branch point on the ancestor timeline is greater than the ancestor's
+# last-record LSN. We had a bug where GetPage incorrectly followed the
+# timeline to the ancestor without waiting for the missing WAL to
+# arrive.
+@pytest.mark.parametrize("remote_storage_kind", available_remote_storages())
+def test_tenant_with_branches_reattach_slow_replay(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_tenant_with_branches_reattach_slow_replay",
+    )
+
+    env = neon_env_builder.init_start()
+    env.pageserver.is_testing_enabled_or_skip()
+    pageserver_http = env.pageserver.http_client()
+
+    # Prepare for the test:
+    #
+    # - Main branch, with a table and two inserts to it.
+    # - A logical replication message between the inserts, so that we can conveniently
+    #   pause the WAL ingestion between the two inserts.
+    # - Child branch, created after the inserts
+    # - Detach the tenant
+    tenant_id, timeline_id = env.neon_cli.create_tenant()
+
+    main_endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
+    with main_endpoint.cursor() as cur:
+        cur.execute("CREATE TABLE test_reattach (t text)")
+        cur.execute("INSERT INTO test_reattach VALUES ('before pause')")
+
+        cur.execute("SELECT pg_logical_emit_message(false, 'neon-test', 'between inserts')")
+
+        cur.execute("INSERT INTO test_reattach VALUES ('after pause')")
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+
+    main_endpoint.stop()
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
+
+    pageserver_http.tenant_detach(tenant_id)
+
+    # This fail point will pause the WAL ingestion on the main branch, after the
+    # the first insert
+    pageserver_http.configure_failpoints([("wal-ingest-logical-message-sleep", "return(5000)")])
```

Comment:
What guarantees that the `main` hasn't already progressed past this? I think it's plausible that it'd be still in flight, but it might also not be.

Reply:
This is probably a duplicate of https://github.com/neondatabase/neon/pull/4170/files#r1190986568

Reply:
Because of the `detach` this works, so everything will be re-walreceived. Somehow I failed to "see the detach".

Perhaps here could be a `remote consistent lsn == 0` assertion? Though, the #4204 might make this much more difficult to hit.

Reply:
Hmm, this does depend on the fact that the layer wasn't flushed and uploaded before the detach happens. If something caused the layers to be uploaded before the detach, or if we detach itself did it, this test would fail to reproduce the bug. The test includes a sanity check that the failpoint is hit after attach, so we would catch that.